File: scripts/normalize.py
import re
import unicodedata


HEBREW_PREFIXES = ("ב", "ל", "מ", "ה", "ו", "כ", "ש")
QUOTE_CHARS = "'\"״׳`´“”‘’"
CONSTRUCT_FORMS = {
    "לשכת": "לשכה",
    "יחידת": "יחידה",
    "חטיבת": "חטיבה",
    "מחלקת": "מחלקה",
    "מפקדת": "מפקדה",
}
PROTECTED_PREFIX_WORDS = set(CONSTRUCT_FORMS) | set(CONSTRUCT_FORMS.values()) | {
    "מודיעין",
    "מבצעים",
    "מפקדה",
    "מחלקה",
    "לוחמה",
    "בסיס",
}
PREFIX_STRIPPABLE_BASES = PROTECTED_PREFIX_WORDS | {
    "אלוף",
    "מבצע",
    "מבצעים",
    "מבצעי",
    "חמל",
    "סייבר",
    "קריפטוגרפיה",
    "הצפנה",
    "צופן",
    "שייטת",
    "סיירת",
    "מטכל",
    "שלדג",
    "דובדבן",
    "מגלן",
    "אגוז",
    "יהלם",
    "לוטם",
    "מצוב",
    "הקריה",
    "נבטים",
    "חצרים",
    "פלמחים",
    "מרפאה",
    "בדיקת",
    "דם",
    "תור",
    "מספרה",
    "ספרייה",
    "משרד",
    "קבלה",
    "אפסנאות",
}


def strip_niqqud(text: str) -> str:
    return "".join(ch for ch in text if unicodedata.category(ch) != "Mn")


def _can_strip_to_known_base(token: str, max_depth: int = 3) -> bool:
    candidate = token
    for _ in range(max_depth):
        if len(candidate) <= 3 or candidate[0] not in HEBREW_PREFIXES:
            return False
        candidate = candidate[1:]
        if candidate in PREFIX_STRIPPABLE_BASES or candidate in CONSTRUCT_FORMS:
            return True
    return False


def normalize_token(token: str, remove_prefix: bool = True) -> str:
    token = strip_niqqud(token).strip().lower()
    token = token.translate(str.maketrans("", "", QUOTE_CHARS))
    token = re.sub(r"[^\u0590-\u05ff0-9a-zA-Z]+", "", token)
    if not token:
        return ""
    if token in CONSTRUCT_FORMS:
        return CONSTRUCT_FORMS[token]
    while remove_prefix and len(token) > 3 and token[0] in HEBREW_PREFIXES:
        candidate = token[1:]
        if (
            candidate not in PREFIX_STRIPPABLE_BASES
            and candidate not in CONSTRUCT_FORMS
            and not _can_strip_to_known_base(token)
        ):
            break
        token = token[1:]
        if token in CONSTRUCT_FORMS:
            return CONSTRUCT_FORMS[token]
    return token


def normalize_text(text: str) -> str:
    text = strip_niqqud(text)
    text = text.translate(str.maketrans({ch: "" for ch in QUOTE_CHARS}))
    text = re.sub(r"[-־–—_/]+", " ", text)
    text = re.sub(r"[^\u0590-\u05ff0-9a-zA-Z\s]+", " ", text)
    tokens = [normalize_token(part) for part in text.split()]
    return " ".join(token for token in tokens if token)


def tokenize(text: str) -> list[str]:
    normalized = normalize_text(text)
    return normalized.split() if normalized else []

File: security-validation-poc/src/core/normalize.ts
// Generic Hebrew normalization — a faithful TypeScript port of scripts/normalize.py,
// rewritten so that the prefix-stripping vocabulary is DATA-DRIVEN (passed in),
// not hard-coded. Server and client construct this with the SAME vocabulary so
// that normalization (and therefore bloom lookups) are deterministic on both sides.

const HEBREW_PREFIXES = new Set(['ב', 'ל', 'מ', 'ה', 'ו', 'כ', 'ש']);

// Hebrew niqqud / cantillation marks + general combining marks.
const NIQQUD_RE = /[\u0591-\u05C7]/g;
const QUOTE_RE = /['"״׳`´\u201C\u201D\u2018\u2019]/g;
const SEPARATOR_RE = /[-־–—_/]+/g;
const KEEP_RE = /[^\u0590-\u05FF0-9a-zA-Z\s]+/g;
const TOKEN_KEEP_RE = /[^\u0590-\u05FF0-9a-zA-Z]+/g;

export interface NormalizationConfig {
  constructForms: Map<string, string>;
  knownBases: Set<string>;
}

export function buildNormalizationConfig(
  constructForms: [string, string][],
  prefixVocabulary: string[],
): NormalizationConfig {
  const cf = new Map(constructForms);
  const bases = new Set(prefixVocabulary);
  // Construct-form targets are themselves valid bases.
  for (const target of cf.values()) bases.add(target);
  return { constructForms: cf, knownBases: bases };
}

export function stripNiqqud(text: string): string {
  return text.normalize('NFD').replace(NIQQUD_RE, '');
}

export function normalizeToken(token: string, cfg: NormalizationConfig): string {
  token = stripNiqqud(token).trim().toLowerCase();
  token = token.replace(QUOTE_RE, '');
  token = token.replace(TOKEN_KEEP_RE, '');
  if (!token) return '';

  if (cfg.constructForms.has(token)) return cfg.constructForms.get(token)!;

  // Conservative, generic prefix stripping: only strip when the remaining
  // candidate is a KNOWN base (from the data), preventing over-stripping.
  while (token.length > 3 && HEBREW_PREFIXES.has(token[0])) {
    const candidate = token.slice(1);
    if (!cfg.knownBases.has(candidate) && !cfg.constructForms.has(candidate)) break;
    token = candidate;
    if (cfg.constructForms.has(token)) return cfg.constructForms.get(token)!;
  }
  return token;
}

export function normalizeText(text: string, cfg: NormalizationConfig): string {
  text = stripNiqqud(text);
  text = text.replace(QUOTE_RE, '');
  text = text.replace(SEPARATOR_RE, ' ');
  text = text.replace(KEEP_RE, ' ');
  const tokens = text
    .split(/\s+/)
    .map((part) => normalizeToken(part, cfg))
    .filter(Boolean);
  return tokens.join(' ');
}

export function tokenize(text: string, cfg: NormalizationConfig): string[] {
  const normalized = normalizeText(text, cfg);
  return normalized ? normalized.split(' ') : [];
}
